PR6717-6734:  CSV_EXPORT_MULTIPLE_FILES and Output filters to spreadsheet export

falken10vdl opened the pull request:
This PR combines both PR6717 and PR6734 in the latest code base.
This allows to have multiple files to export to CSV and also the addition of "output filters" which allow to filter the information that it is exported to csv.

Cheers!

falken10vdl commented:
@steverugi Here are the PR6717 and PR6734 in the latest code base.

<img width="3827" height="1840" alt="image" src="https://github.com/user-attachments/assets/a145fc49-f7db-42ca-82f5-28c771bf6f71" />

This is merged in the latest bonsaiPR: https://github.com/falken10vdl/bonsaiPR/releases/tag/v0.8.4-alpha251028
In case you want to test it.
Cheers!

theoryshaw commented:
I might be using this wrong, but I can't seem to name an .ods export, it seems to force the name `combined.ods`

https://github.com/user-attachments/assets/dd5795f8-6366-4292-b8ee-35fea8272711

...

On another note, the combined.ods doesn't seem to open correctly in Librecalc.

https://github.com/user-attachments/assets/096ec013-37fe-4b1d-9eb4-d10cd094eb20

falken10vdl commented:
@theoryshaw You are completely right! My bad... I just rebased and solved for some conflict in webui. However I have seen that the export has also changed. 
Now I generate a pandas frame (the one that then goes to webUI) and convert this one to the selected file format.
I have also added as suggested filename the concatenation of the files that are selected for exporting.

https://github.com/user-attachments/assets/2e09075b-5e73-462a-a289-bfa5763b7577

The bonsaiPR addon has been regenerated:
https://github.com/falken10vdl/bonsaiPR/releases/tag/v0.8.4-alpha251028

Cheers!

theoryshaw commented:
Hi. I can't seem to get 'import' to work.

falken10vdl commented:
I did not touch the import process but now I realize that since in the export I add a column with the filename to distinguish possible identical rows from different files, clearly the import process needs to be updated to loop over the files.
Let me fix this.
Thanks!

falken10vdl commented:
@theoryshaw I have made updates to properly import taking in account the first column as filename.
Thanks for the feedback!

theoryshaw commented:
Import seems to hang for a long time.  It does stop eventually, but the import does not work. 
files: https://hub.openingdesign.com/OpeningDesign/Cool_Car_Guys/src/commit/da91412469158796991123b456624b4c5392b053/Open/Models/Bonsai

https://github.com/user-attachments/assets/3dbaaa4a-a774-410a-bd36-1b46154a445c

falken10vdl commented:
@theoryshaw I agree it is quite slow. I am doing some tests but I see that we could build up a dictionary instead of going row by row looking for the globalId in the whole file.
However are we sure that the import logic is able to handle attributes with group different than "none" (@maxfb87 @steverugi do you agree?)? I mean that if you start to group, the output will have "fake global Ids". If you try your example by exporting first without grouping (and importing without count, as that one is not a valid attribute to store info) I do not see any problem to import data.
To summarize we need to tackle the following:
- If attributes are "grouped" -> maybe in import just report back to the user that this is not a valid importing workflow and stop?
- Speed. I think that a dictionary will improve. Maybe also add some sort of "progress" feedback (so far I have not see in bonsai progress toolbars. Only the funny cursor changing doing file load)

Cheers!

steverugi commented:
@falken10vdl sorry I missed a few episodes here and there, need to refresh my memory, will check soon and revert, cheers

maxfb87 commented:
@falken10vdl i wish i could help but unfortunately at the moment i'm not focus on that...if i can help i'll tell you

theoryshaw commented:
@falken10vdl Is there any new plumbing in the PR, that increases the import time?  
It goes pretty fast on the main branch.

theoryshaw commented:
Forgot to mention, I added `count` and `material` as skipping patterns a couple days ago...
https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.8.0/src/ifccsv/ifccsv.py#L513

falken10vdl commented:
I have completely reviewed the process of importing and changed the import to be a python generator. In this way we can switch the import operator to modal and show a progress bar. I have arbitrarily set the refresh rate of the modal to be 50ms so the UI is responsive but still the import process is fast enough... 
Please take a look:

https://github.com/user-attachments/assets/cda8b647-80b0-4696-8490-dc2d9ecd7721

Cheers!

steverugi commented:
thank you @falken10vdl , you are a star!
I've yet to install your PR but looks very good in the video already

falken10vdl commented:
thanks @steverugi  I am going to rebase ti since it looks there are have been some changes in the code base lately.
Eventually I guess that the handling of multiple files should be done by a project management tool. I am taking a look to the opencdeserver and hope to have something to share in the near future in which we can handle multiple ifc files (and their drawings/maerials/etc) within a project.

falken10vdl commented:
@theoryshaw can you please try it? You always have good tests to clean up ;) Thanks!

steverugi commented:
@falken10vdl I tried latest windows PR 251207 and looks OK for exporting from multiple files (including linked ones), do you think it's possible to include the model in memory too? it's not a big issue since it can be loaded from the file.
thanks a lot for your work, truly amazing

PS the UI doesn't seem to work 100%, column-slab extension Shift-E fails for instance, are you aware of it?
EDIT: it's not because of your PR, also the 'unstable' does the same, will post an issue

steverugi commented:
> thanks @steverugi I am going to rebase ti since it looks there are have been some changes in the code base lately. Eventually I guess that the handling of multiple files should be done by a project management tool. I am taking a look to the opencdeserver and hope to have something to share in the near future in which we can handle multiple ifc files (and their drawings/maerials/etc) within a project.

you mean something like web dashboard callable from Bonsai to handle federated models? that's cool

falken10vdl commented:
> > thanks @steverugi I am going to rebase ti since it looks there are have been some changes in the code base lately. Eventually I guess that the handling of multiple files should be done by a project management tool. I am taking a look to the opencdeserver and hope to have something to share in the near future in which we can handle multiple ifc files (and their drawings/maerials/etc) within a project.
> 
> you mean something like web dashboard callable from Bonsai to handle federated models? that's cool

@steverugi 
I do not know if it makes more sense to do an extra blender addon to handle a project with several files or better to use the opencde (probably integrated with forgejo) and add that functionality there (probably this makes more sense) i ma tryign to learn a little bit how the opencd, bcf, document api, etc is linked together because that would be the proper way to handle several files within a higher hierarchy I suppose.
What do you think?

falken10vdl committed:
CSV export multiple files and output filters with imperial unit support

- Add support for exporting multiple IFC files with combined output
- Implement output filter groups with comparison operators (=, !=, >, <, >=, <=, *=, termons=, regex)
- Add attribute type system (string, float, int, bool, imperial_string) for type-safe filtering
- Implement imperial unit parsing and formatting (supports formats like 29' 2'', 8' 10 3/16")
- Add incremental CSV import with modal progress tracking
- Support export to CSV, XLSX, ODS, and web UI formats
- Add IFC file selection panel with linked files support
- Preserve settings in JSON templates
- Add comprehensive documentation with images

falken10vdl committed:
Remove outdated documentation files

falken10vdl committed:
Add data type support to CSV attributes and enhance output filter functionality

- Introduced a new `data_type` property for `CsvAttribute` to specify the type of attribute for filtering and export.
- Updated comparison operators in `CsvOutputFilter` to include descriptions for approximate equality.
- Added `filter_mode` property to `CsvOutputFilter` for managing how filters are applied (Add, Subtract, Filter).
- Implemented `ToggleOutputFilterInclusion` operator to cycle through filter modes in the UI.
- Enhanced UI to display the new data type property and filter mode options.

falken10vdl committed:
Refactor CSV export to include filename and global ID, update UI and properties accordingly

falken10vdl committed:
Refactor CSV export logic to use generator and cleanup

steverugi commented:
Hi @falken10vdl 
what do you think it takes to pull this in the unstable release?
I know (and much appreciate) it's in yours but I was wondering when this feature will be more 'official'
thanks

falken10vdl commented:
@steverugi 
Have you been able to test it in depth? Any thing seen? Both import/export are working fine for you? It is a major PR.

steverugi commented:
OK @falken10vdl will run some test and revert, thanks

falken10vdl commented:
@steverugi I will check in my side what are the tests already defined in the code to pass them or adapt them in the PR as I am sure that the need to add the filename to differentiate the multiple files needs to be addressed in the automated testing of the release.

falken10vdl commented:
@steverugi I think we need to add a "file quantity" field in the export to cope for cases like when you have multiple times the same file linked (https://github.com/IfcOpenShell/IfcOpenShell/pull/7570). Ex. If a site has 50 equal buildings linked (50 links with same filename) it foes not make sense to export it 50 times.
Cheers!

steverugi commented:
> @steverugi I think we need to add a "file quantity" field in the export to cope for cases like when you have multiple times the same file linked (#7570). Ex. If a site has 50 equal buildings linked (50 links with same filename) it foes not make sense to export it 50 times. Cheers!

Good idea, thanks

theoryshaw commented:
Converted to a draft, as it currently conflicts with `v0.8.0`.

see https://github.com/falken10vdl/bonsaiPR/releases/tag/v0.8.5-alpha2603141648

please note the `First detected failing` date is most likely wrong, since that was the date this _automated conflict report_ functionality was 'turned on'.

Conflict Report:

-   [**PR #7295**](https://github.com/IfcOpenShell/IfcOpenShell/pull/7295): PR6717-6734: CSV_EXPORT_MULTIPLE_FILES and Output filters to spreadsheet export
    -   Reason: Fails to merge against base (problem with PR itself)
    -   First detected failing: 2026-03-10
    -   Base commit at first detection:  `972a1fd3091d172de30c539c6423e87f5a3ac2db`
    -   Conflicting files:  `src/bonsai/bonsai/bim/module/csv/operator.py`,  `src/bonsai/bonsai/bim/module/csv/ui.py`
    -   Possible breaking commits:  `ff3933a11 Remove some unused imports`,  `953c5fddc bonsai - sort imports`,  `fe130ebb5 typing`,  `8b4683aef replace api.run with static methods`,  `4593e5aea Fix ExportHelper not optional filename_ext #6463 (4594f59)`
